e2etests/tests/graceful_restart.go

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package tests
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/openshift-kni/k8sreporter"
+	"go.universe.tf/e2etest/pkg/frr/container"
+
+	frrk8sv1beta1 "github.com/metallb/frr-k8s/api/v1beta1"
+	"github.com/metallb/frrk8stests/pkg/config"
+	"github.com/metallb/frrk8stests/pkg/dump"
+	"github.com/metallb/frrk8stests/pkg/infra"
+	"github.com/metallb/frrk8stests/pkg/k8s"
+	"github.com/metallb/frrk8stests/pkg/k8sclient"
+	"github.com/metallb/frrk8stests/pkg/routes"
+	. "github.com/onsi/gomega"
+	frrconfig "go.universe.tf/e2etest/pkg/frr/config"
+	"go.universe.tf/e2etest/pkg/ipfamily"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+var _ = ginkgo.Describe("Establish BGP session with EnableGracefulRestart", func() {
+	var (
+		cs       clientset.Interface
+		updater  *config.Updater
+		reporter *k8sreporter.KubernetesReporter
+		nodes    []corev1.Node
+	)
+
+	cleanup := func(u *config.Updater) error {
+		for _, c := range infra.FRRContainers {
+			if err := c.UpdateBGPConfigFile(frrconfig.Empty); err != nil {
+				return fmt.Errorf("clear config in the infra container failed: %w", err)
+			}
+		}
+		if err := u.Clean(); err != nil {
+			return fmt.Errorf("clear config in the API failed: %w", err)
+		}
+		return nil
+	}
+
+	ginkgo.BeforeEach(func() {
+		var err error
+
+		reporter = dump.NewK8sReporter(k8s.FRRK8sNamespace)
+		updater, err = config.NewUpdater()
+		Expect(err).NotTo(HaveOccurred())
+
+		err = cleanup(updater)
+		Expect(err).NotTo(HaveOccurred(), "cleanup config in API and infra containers")
+
+		cs = k8sclient.New()
+		nodes, err = k8s.Nodes(cs)
+		Expect(err).NotTo(HaveOccurred())
+
+		h, err := k8s.FRRK8isDaemonSetReady(cs)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(h).Should(BeTrue(), "frr-k8s daemonset should be ready before test")
+	})
+
+	ginkgo.AfterEach(func() {
+		if ginkgo.CurrentSpecReport().Failed() {
+			testName := ginkgo.CurrentSpecReport().LeafNodeText
+			dump.K8sInfo(testName, reporter)
+			dump.BGPInfo(testName, infra.FRRContainers, cs)
+		}
+	})
+
+	ginkgo.Context("When restarting the frrk8s deamon pods", func() {
+
+		assertFuncPerEntry := func(ipFam ipfamily.Family, prefix string) {
```

Comment:
not blocking, leaving only because we might change the other comments, but: can we embed this as part of the table entry? It's gonna be as per all the other table tests and we are not really reusing the function

Reply:
done (but imo is less readable that way)